<launch>

<node pkg="sr_slam" type="sr_main_plane" name="sr_slam_plane" output="log">

  <!-- sr4k data parameters -->
  <!-- param name="sr_data_file_dir"              value="/home/davidz/work/data/SwissRanger4000/with_gt/dataset_31" /-->
  <param name="sr_data_file_dir"              value="/home/davidz/work/data/SwissRanger4000/bdat/dataset_3" />
  <param name="sr_data_file_dir"              value="/home/davidz/work/data/SwissRanger4000/exp/dataset_70" />
  <param name="sr_data_file_dir"              value="/home/davidz/work/data/SwissRanger4000/exp/dataset_82" />
  <param name="sr_data_file_dir"              value="/home/davidz/work/data/SwissRanger4000/exp/dataset_123" />

  <param name="sr_data_prefix"                value="d1"/>
  <param name="sr_data_suffix"                value="bdat"/> <!-- dat -->
  
  <!-- syn_start_frame: dataset_3: 12 - 1100, dataset_2: 2 - 2310, dataset_1: 3-1200 -->
  <param name="sr_start_frame"                value="3"/> <!-- notice: for matching timestamp in ground_truth, start should be 1 -->
  <param name="sr_end_frame"                  value="880"/> 
  <param name="sr_new_file_version"           value="false"/> <!-- true -->

  <param name="config/b_add_floor_into_graph"   value="true"/> <!-- true -->
  <param name="config/fixed_frame_name"       value="/map"/>
  <param name="config/base_frame_name"        value=""/>
  <param name="config/bagfile_name"           value=""/>

  <!-- display style -->
  <param name="config/cloud_display_type"             value="POINTS"  />

  <!-- write nodes into disk -->
  <param name="config/save_node_path"                 value="/home/davidz/work/data/SwissRanger4000/vro_results/bdat/dataset_3" />
  <param name="config/process_node_method"            value="slam" /> <!-- slam, write, display  -->
  <param name="config/read_node_from_disk"            value="false" /> <!--  -->

  <!-- these two parameters control the performance of offline slam  -->
  <param name="config/concurrent_edge_construction"   value="true"/> <!-- weather use QTthreads to run comparison in parallel -->
  <param name="run_slam_offline"              value="false"/> <!-- in offline model, do not take motion speed into consideration-->
  <param name="b_downsample_pc"               value="false"/> <!-- weather to downsample the point cloud -->
  <param name="filter_voxel_size"             value="0.02"/> <!-- voxel grid size -->

  <!-- keep good nodes, try to maintain the process if vro fails -->
  <param name="config/keep_good_nodes"        value="true"/>

  <!-- display the 3D map, and reproject 2D point -->
  <param name="config/store_pointclouds"      value="true"/> 
  <param name="config/visualize_keyframes_only" value="true"/>
  <param name="config/glwidget_without_clouds" value="false"/>
  <param name="config/send_pc2_vis"           value="true"/>

  <!-- Visual Features, to activate GPU-based features see CMakeLists.txt -->
  <param name="config/feature_detector_type"         value="SIFTGPU"/><!-- If SIFTGPU is enabled in CMakeLists.txt, use SURF here -->
  <param name="config/feature_extractor_type"        value="SIFTGPU"/><!-- If SIFTGPU is enabled in CMakeLists.txt, use SURF here -->
  <param name="config/min_keypoints"                 value="50"/><!-- Extract no less than this many ... -->
  <param name="config/min_matches"                   value="10"/> <!-- Minimal number of matched features before RANSAC 29? -->
  <param name="config/nn_distance_ratio"             value="0.95"/> <!-- 0.5 for SIFT, 0.95 for SIFTGPU Feature correspondence is valid if distance to nearest neighbour is smaller than this parameter times the distance to the 2nd neighbour. This needs to be 0.9-1.0 for SIFTGPU w/ FLANN, since SIFTGPU Features are normalized-->
  
   <!-- decide detect obstacles in the front  -->
  <param name="config/b_front_obstacle_detect"     value="true"/> <!-- check some obstacle appear in the front -->
  <param name="config/obstacle_min_z"              value="-0.5"/>
  <param name="config/obstacle_len_y"              value="0.4" />
  <param name="config/obstacle_max_x"              value="3" />
  <param name="config/obstacle_density_threshold"  value="0.04" />
  <param name="config/obstacle_number_pt_threshold"  value="1000" />

  <!-- max rotation and translation for a VO match between two frames -->
  <param name="config/max_rotation_degree"          value="90"/> 
  <param name="config/max_translation_meter"        value="1.5"/>

  <param name="config/max_dist_for_inliers"         value="0.05" /> <!-- 6 3 "Mahalanobis distance for matches to be considered inliers by ransac"-->
  
  <!-- vro strategy -->
  <param name="config/vro_strategy"                 value="vro_my" /> <!-- " VRO strategy: VRO_ori, VRO_my, VRO_plane, VRO_plane_em" -->
  <param name="config/ransac_iterations"            value="2000" /> <!-- "Number of iterations for registration" -->
  <param name="config/b_ransac_fixed_iter"           value="false"/> <!-- weather to use fixed   -->

  <!-- sift and siftgpu parameters -->
  <param name="sift_num_features"                   value="0" /> <!-- number of returned sift features -->
  <param name="sift_octave_layers"                  value="5" /> <!-- number octave layers -->
  <param name="sift_contrast_threshold"             value="0.04" /> <!-- initial threshold for keypint detection -->
  <param name="siftgpu_contrast_threshold"          value="0.0008" /> <!-- 0.0018 initial DOG threshold for siftgpu -->
  <param name="siftgpu_edge_threshold"              value="10." /> <!-- edge threshold for siftgpu  -->

  <!-- Algortithm settings -->
  <param name="config/optimizer_skip_step"    value="20" /><!-- 20 Optimize every n-th frame. Set negative for offline operation -->
  <param name="config/min_translation_meter"         value="0.05"/><!-- 0.05 frames with motion less than this, will be omitted -->
  <param name="config/min_rotation_degree"           value="2"/><!-- frames with motion less than this, will be omitted -->
  <param name="config/predecessor_candidates"        value="5"/><!-- 5 search through this many immediate predecessor nodes for corrspondences -->
  <param name="config/neighbor_candidates"           value="0"/><!-- 5 search through this many graph neighbour nodes for corrspondences -->
  <param name="config/min_sampled_candidates"        value="0"/><!-- 5 search through this many uniformly sampled nodes for corrspondences -->
  <param name="config/max_connections"               value="5"/><!-- 10 -1 stop after this many succesfully found spation relations -->
</node>


</launch>
